In [ ]:
print('Welcome to IE507_Lab2_Ex1_Solution!!')

Welcome to IE507_Lab2_Ex1_Solution!!


In [ ]:
!pip install -q pyomo

     |████████████████████████████████| 9.0MB 2.5MB/s 
     |████████████████████████████████| 51kB 6.8MB/s 
     |████████████████████████████████| 256kB 45.7MB/s 
     |████████████████████████████████| 163kB 50.9MB/s 


In [ ]:
from pyomo.environ import *

In [ ]:
import numpy as np

In [ ]:
# create a model
model = ConcreteModel()

In [ ]:
obj_coef_c = np.array([-10.1, 20.15, 60.5, 80.05, 0.15, 40.5, -0.05, -30.02, 40.65, 0.125])

In [ ]:
constr1_coef_s = np.array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1])
constr2_coef_w = np.array([0, 1.6, -2.02, 0, 1.01, -2.005, 0, 1.39, 1, -1.214])

#declaring RHS of the constraints as variables
constr1_rhs_b1 = 151.2
constr2_rhs_b2 = 44.6

In [ ]:
#setting the lower bound and upper bounds for the variables
lower_bound = np.array([-np.inf, 1, 0, 0, 7, 0,-np.inf, -np.inf, 1, 1])
upper_bound = np.array([4, 3, np.inf, 2, 10, np.inf, 13, 20, np.inf, 21])

In [ ]:
# set of column indices: since we have N=10 variables, we can take column indices to be the list [0,1,2,...,10]
col_indices = np.arange(10)
print(col_indices)

[0 1 2 3 4 5 6 7 8 9]


In [ ]:
#declaration of the decision variables in the model.
model.x = Var(col_indices)

In [ ]:
# creation a ConstraintList to hold multiple constraints
model.constraints = ConstraintList()

In [ ]:
#add model constraints one by one to the list 
#first adding the constraint s1*x1+ s2*x2+...+ sN*xN <= b1 
model.constraints.add(sum(constr1_coef_s[j]*model.x[j] for j in col_indices) <= constr1_rhs_b1)

In [ ]:
#subsequently adding the constraint w1*x1 + w2*x2 + ...+  wN*xN = b2
model.constraints.add(sum(constr2_coef_w[j]*model.x[j] for j in col_indices) == constr2_rhs_b2)

In [ ]:
#adding the upper bound and lower bounds for the variables
for j in col_indices:
  model.x[j].setlb(lower_bound[j])
  model.x[j].setub(upper_bound[j])

In [ ]:
#adding the model objective
model.objective = Objective(expr = sum(obj_coef_c[j]*model.x[j] for j in col_indices), sense=minimize)

In [ ]:
model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    3 : {1, 2, 3}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   10 : {0, 1, 2, 3, 4, 5, 6, 7, 8, 9}

1 Var Declarations
    x : Size=10, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :  -inf :  None :   4.0 : False :  True :  Reals
          1 :   1.0 :  None :   3.0 : False :  True :  Reals
          2 :   0.0 :  None :   inf : False :  True :  Reals
          3 :   0.0 :  None :   2.0 : False :  True :  Reals
          4 :   7.0 :  None :  10.0 : False :  True :  Reals
          5 :   0.0 :  None :   inf : False :  True :  Reals
          6 :  -inf :  None :  13.0 : False :  True :  Reals
          7 :  -inf :  None :  20.0 : False :  True :  Reals
          8 :   1.0 :  None :   inf : False :  True :  Rea

In [ ]:
!apt-get install -y -qq glpk-utils

Selecting previously unselected package libsuitesparseconfig5:amd64.
(Reading database ... 144556 files and directories currently installed.)
Preparing to unpack .../libsuitesparseconfig5_1%3a5.1.2-2_amd64.deb ...
Unpacking libsuitesparseconfig5:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libamd2:amd64.
Preparing to unpack .../libamd2_1%3a5.1.2-2_amd64.deb ...
Unpacking libamd2:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libcolamd2:amd64.
Preparing to unpack .../libcolamd2_1%3a5.1.2-2_amd64.deb ...
Unpacking libcolamd2:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libglpk40:amd64.
Preparing to unpack .../libglpk40_4.65-1_amd64.deb ...
Unpacking libglpk40:amd64 (4.65-1) ...
Selecting previously unselected package glpk-utils.
Preparing to unpack .../glpk-utils_4.65-1_amd64.deb ...
Unpacking glpk-utils (4.65-1) ...
Setting up libsuitesparseconfig5:amd64 (1:5.1.2-2) ...
Setting up libcolamd2:amd64 (1:5.1.2-2) ...
Setting up libamd2:amd64 

In [ ]:
SolverFactory('glpk', executable='/usr/bin/glpsol').solve(model).write()

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: -452.7909
  Upper bound: -452.7909
  Number of objectives: 1
  Number of constraints: 4
  Number of variables: 11
  Number of nonzeros: 28
  Sense: minimize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: optimal
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.010019779205322266
# ----------------------------------------------------------
#   Solution Information
# --------------------------

In [ ]:
# display solution
print('\nObjective = ', model.objective())

print('\nDecision Variables')
for i in col_indices:
  print('x[',i,'] = ', model.x[i].value)

print('\nConstraints')
model.constraints.display()


Objective =  -452.7909

Decision Variables
x[ 0 ] =  4.0
x[ 1 ] =  3.0
x[ 2 ] =  0.0
x[ 3 ] =  0.0
x[ 4 ] =  10.0
x[ 5 ] =  0.0
x[ 6 ] =  13.0
x[ 7 ] =  20.0
x[ 8 ] =  3.114
x[ 9 ] =  1.0

Constraints
constraints : Size=3
    Key : Lower : Body               : Upper
      1 :  None :             54.114 : 151.2
      2 :  None :             54.114 : 151.2
      3 :  44.6 : 44.599999999999994 :  44.6


The optimal solution and the vale of objective function are:
x1=4.0,x2 =  3.0,
x3 =  0.0,
x4 =  0.0,
x5 =  10.0,
x6 =  0.0,
x7 =  13.0,
x8 =  20.0,
x9=  3.114,
x10 =  1.0
and Objective = -452.7909